### PR TITLE
Sp 142 configure

### PR DESCRIPTION
**Pull request by apurvapatkar**
- Update Configure, Scan, and EndScan commands on SubArrayNode to receive/send JSON for Dish configuration (support for SP-142) 
- Update Configure, Scan, and EndScan commands on DishLeafNode to receive/send JSON for Dish configuration  (support for SP-142)
- Update Configure, Scan, and EndScan commands on DishMaster to receive JSON for Dish configuration (support for SP-142)

### CODE REVIEW

File: tmcprototype/CspSubarray/CspSubarray/CspSubarray.py [hunk 1]
```diff
@@ -127,6 +128,8 @@ def init_device(self):
         self._receptor = []
         self._correlations = " "
         self._health_state = 0
```
Comment (mangeshska) [inline tmcprototype/CspSubarray/CspSubarray/CspSubarray.py]: self._health_state = 0
self._obs_state = 1
self._opstate = 2
Not clear which state is represented by '2'. Use of enum recommended. Else add comment for better understanding.

Comment (jyotinpo) [reply tmcprototype/CspSubarray/CspSubarray/CspSubarray.py]: Nice suggestion, this feature is available in PyTango 9.2.5+. We need to test it before implementing throughout the project. It was implemented as part of [this](https://github.com/tango-controls/pytango/pull/221) pull request in PyTango.

Comment (JayantKumbhar) [reply tmcprototype/CspSubarray/CspSubarray/CspSubarray.py]: 1. Resoved these comments in commit [aaf968f95e7ba3fb5e6aed85ed5e3ef9ff6bc2cf].
2. Added comments on CspSubarray.
3. Added ENUMs on CspMasterLeafNode.

Comment (apurvapatkar) [reply tmcprototype/CspSubarray/CspSubarray/CspSubarray.py]: Updated the SubarrayNode and CentralNode accordingly. The commit id is: [0938403d20f73f4fdc2a112cbd901b51afca5fe3](https://github.com/ska-telescope/tmc-prototype/commit/0938403d20f73f4fdc2a112cbd901b51afca5fe3)

Comment (kalyanithigale) [reply tmcprototype/CspSubarray/CspSubarray/CspSubarray.py]: Resolved comments for DishLeafNode and SdpSubarrayLeafNode in commit  ba85d10655f89420768871fc02d63fd651de5e55

File: tmcprototype/CspSubarray/CspSubarray/CspSubarray.py [hunk 2]
```diff
@@ -214,7 +217,10 @@ def read_correlationDigitalBandpassCorrectionCoefficients(self):
     @DebugIt()
     def ConfigureScan(self, argin):
         # PROTECTED REGION ID(CspSubarray.ConfigureScan) ENABLED START #
-        pass
+        self._obs_state = 1                 # Set ObsState to CONFIGURING
+        print("ConfigureScan is invoked successfully on CspSubarray. Argin:", argin)
```
Comment (jyotinpo) [inline tmcprototype/CspSubarray/CspSubarray/CspSubarray.py]: Remove unnecessary print statements.

File: tmcprototype/CspSubarrayLeafNode/CspSubarrayLeafNode/CONST.py [hunk 3]
```diff
@@ -2,52 +2,41 @@
...
-STR_FAILED = " :-> FAILED"
-STR_UNKNOWN = " :-> UNKNOWN"
 STR_ERR_MSG = "Error message is: "
-STR_INIT_SUCCESS = "CspSubarrayLeafNode is initialized successfully."
-STR_FALSE = "False"
-STR_ASSIGN_RESOURCES_SUCCESS = "Resources are assigned successfully"
-STR_RELEASE_ALL_RESOURCES_SUCCESS = "All resources are removed successfully"
-STR_CSP_CBF_DEV_NAME = "mid_csp_cbf/sub_elt/master"
+STR_ASSIGN_RESOURCES_SUCCESS = "Resources are assigned successfully on CspSubarray."
+STR_RELEASE_ALL_RESOURCES_SUCCESS = "All resources assigned to CSP Subarray are removed successfully"
```
Comment (jyotinpo) [inline tmcprototype/CspSubarrayLeafNode/CspSubarrayLeafNode/CONST.py]: Can change it to STR_REMOVE_ALL_RECEPTORS_SUCCESS.

File: tmcprototype/CspSubarrayLeafNode/test/CspSubarrayLeafNode_test.py [hunk 4]
```diff
@@ -91,11 +91,29 @@ def test_Reset(self, create_cspsubarray1_proxy):
...
-    #     """Test for ConfigureScan"""
-    #     # PROTECTED REGION ID(CspSubarrayLeafNode.test_ConfigureScan) ENABLED START #
-    #     self.device.ConfigureScan("")
-    #     # PROTECTED REGION END #    //  CspSubarrayLeafNode.test_ConfigureScan
+    def test_ConfigureScan(self, tango_context):
+        """Test for ConfigureScan"""
+        # PROTECTED REGION ID(CspSubarrayLeafNode.test_ConfigureScan) ENABLED START #
+        test_input = ''
+        test_input = '{"dish":{"receptorIDList":["0001","0002"]}}'
+        print("test_input", test_input)
```
Comment (jyotinpo) [inline tmcprototype/CspSubarrayLeafNode/test/CspSubarrayLeafNode_test.py]: Remove unnecessary print statements.

File: tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py [hunk 5]
```diff
@@ -270,7 +271,8 @@ def convert_radec_to_azel(self, data):
 
             # Compute Target Coordinates
             target_radec = data[0]
-            desired_target = katpoint.Target(target_radec)
+            print ("target_radec:", target_radec)
```
Comment (jyotinpo) [inline tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py]: Can this print be removed?

File: tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py [hunk 6]
```diff
@@ -305,6 +307,7 @@ def convert_radec_to_azel(self, data):
             self.RaDec_AzEl_Conversion = True
         except ValueError as value_err:
             print(CONST.ERR_RADEC_TO_AZEL_VAL_ERR)
+            print(value_err)
```
Comment (jyotinpo) [inline tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py]: Remove unused print statements.

File: tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py [hunk 7]
```diff
@@ -593,12 +601,18 @@ def EndScan(self, argin):
...
         """
         excpt_count = 0
         excpt_msg = []
+        # TODO: Accept EndScan argument in JSON format
+        # jsonArgument = json.loads(argin)
+        # timestamp = jsonArgument['timestamp']
+        # print("End Scan timestamp:", timestamp)
         try:
             if type(float(argin)) == float:
+                print ("Sending end scan command...")
```
Comment (jyotinpo) [inline tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py]: Remove print statement.

File: tmcprototype/CspSubarrayLeafNode/CspSubarrayLeafNode/CONST.py [hunk 8]
```diff
@@ -2,52 +2,41 @@
...
-STR_DEGRADED = " :-> DEGRADED"
-STR_FAILED = " :-> FAILED"
-STR_UNKNOWN = " :-> UNKNOWN"
 STR_ERR_MSG = "Error message is: "
-STR_INIT_SUCCESS = "CspSubarrayLeafNode is initialized successfully."
-STR_FALSE = "False"
-STR_ASSIGN_RESOURCES_SUCCESS = "Resources are assigned successfully"
-STR_RELEASE_ALL_RESOURCES_SUCCESS = "All resources are removed successfully"
-STR_CSP_CBF_DEV_NAME = "mid_csp_cbf/sub_elt/master"
+STR_ASSIGN_RESOURCES_SUCCESS = "Resources are assigned successfully on CspSubarray."
```
[the remark this answers is not in the document]
Comment (JayantKumbhar) [reply tmcprototype/CspSubarrayLeafNode/CspSubarrayLeafNode/CONST.py]: Changed STR_ASSIGN_RESOURCES_SUCCESS to STR_ADD_RECEPTORS_SUCCESS in CONST.py file of CspSubarrayLeafNode.
Commit ID: [40fecece84537688971ab8a4fe9900f29f02b21b]

[the remark this answers is not in the document]
Comment (JayantKumbhar) [reply tmcprototype/CspSubarrayLeafNode/CspSubarrayLeafNode/CONST.py]: Changed STR_ASSIGN_RESOURCES_SUCCESS to STR_ADD_RECEPTORS_SUCCESS in CONST file of CspSubarrayLeafNode.
Commit Id: [40fecece84537688971ab8a4fe9900f29f02b21b]

File: tmcprototype/CspSubarrayLeafNode/test/CspSubarrayLeafNode_test.py (hunk 4 above)
Comment (JayantKumbhar) [reply tmcprototype/CspSubarrayLeafNode/test/CspSubarrayLeafNode_test.py]: Removed.
Commit Id: [40fecec]

File: tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py (hunk 5 above)
Comment (kalyanithigale) [reply tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py]: Print statement removed. Commit: 1cb8e394dcea0d91b6be489435940b602451797d

File: tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py (hunk 6 above)
Comment (kalyanithigale) [reply tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py]: Removed. Commit: 1cb8e394dcea0d91b6be489435940b602451797d

File: tmcprototype/CspSubarrayLeafNode/CspSubarrayLeafNode/CONST.py (hunk 3 above)
Comment (JayantKumbhar) [reply tmcprototype/CspSubarrayLeafNode/CspSubarrayLeafNode/CONST.py]: Changed STR_RELEASE_ALL_RESOURCES_SUCCESS to STR_REMOVE_ALL_RECEPTORS_SUCCESS in CONST file of CspSubarrayLeafNode.
Commit Id: [40fecec]

File: tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py (hunk 7 above)
Comment (kalyanithigale) [reply tmcprototype/DishLeafNode/DishLeafNode/DishLeafNode.py]: Removed. Commit: 1cb8e394dcea0d91b6be489435940b602451797d

**Comment by apurvapatkar** [comment]
- Update Configure, Scan, and EndScan commands on SubArrayNode to receive/send JSON for Dish configuration (support for SP-142)
- Update Configure command on DishLeafNode to receive/send JSON for Dish configuration (support for SP-142)
- Update Configure command on DishMaster to receive JSON for Dish configuration (support for SP-142)

**Pull request merged**